class.jetpack-signature.php

```diff
@@ -230,12 +241,27 @@ function normalized_query_parameters( $query_string ) {
 		return $pairs;
 	}
 
-	function encode_3986( $string ) {
-		$string = rawurlencode( $string );
-		return str_replace( '%7E', '~', $string ); // prior to PHP 5.3, rawurlencode was RFC 1738
+	function encode_3986( $string_or_array ) {
+		if ( is_array( $string_or_array ) ) {
+			$encoded = array();
+			foreach ( $string_or_array as $key => $value) {
+				$encoded[ $key ] = $this->encode_3986( $value );
+			}
+			return $encoded;
+		}
+
+		$string_or_array = rawurlencode( $string_or_array );
+		return str_replace( '%7E', '~', $string_or_array ); // prior to PHP 5.3, rawurlencode was RFC 1738
 	}
 
 	function join_with_equal_sign( $name, $value ) {
+		if ( is_array( $value ) ) {
+			$result = array();
+			foreach ( $value as $array_key => $array_value ) {
+				$result[] = $name . '[' . $array_key . ']' . '=' . $array_value;
+			}
+			return $result;
```

Comment:
Can we use [`add_query_arg`](https://developer.wordpress.org/reference/functions/add_query_arg/) (with an empty string for the URL param) instead of the `foreach`?

Or if we can't, maybe at least `array_reduce`? :slightly_smiling_face:

Reply:
We can't use `add_query_arg` with an empty string for URL, because in that case it'll use the current URL.

We can't use `array_reduce` in a beautiful way, because here in Jetpack we're limited by using PHP 5.2, and it doesn't support anonymous functions, and the `array_reduce` callback would need to have access to the upper level field name...

So, we could go with a functional approach, but it wouldn't be more readable than what we've got here IMHO.

Reply:
Fair enough.